src/tpm2_swtpm.c

```diff
@@ -0,0 +1,289 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1335, USA
+ */
+
+
+
+/**
+ * This implements a subset of TPM TCP protocol as described in
+ * "TPM-Rev-2.0-Part-4-Supporting-Routines-01.38-code"
+ *
+ * This is intended for testing with a simulator such as
+ * http://ibmswtpm.sourceforge.net/ or
+ * https://github.com/stefanberger/swtpm
+ *
+ * See docs/SWTPM.md
+ */
+
+#ifdef WOLFTPM_SWTPM
+#include <wolftpm/tpm2.h>
+#include <wolftpm/tpm2_swtpm.h>
+#include <wolftpm/tpm2_packet.h>
+
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <stdio.h>
+
+#include <sys/types.h>
+#include <sys/socket.h>
+#include <netdb.h>
+
+
+#ifndef TPM2_SWTPM_HOST
+#define TPM2_SWTPM_HOST         "localhost"
+#endif
+#ifndef TPM2_SWTPM_PORT
+#define TPM2_SWTPM_PORT         "2321"
+#endif
+
+static TPM_RC SwTpmTransmit(TPM2_CTX* ctx, const void* buffer, ssize_t bufSz)
+{
+    TPM_RC rc = TPM_RC_SUCCESS;
+    ssize_t wrc = 0;
+
+    if (ctx == NULL || ctx->tcpCtx.fd <= 0 || buffer == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    wrc = write(ctx->tcpCtx.fd, buffer, bufSz);
+    if (bufSz != wrc) {
+        rc = SOCKET_ERROR_E;
+    }
+
+#ifdef WOLFTPM_DEBUG_VERBOSE
+    if (wrc < 0) {
+        printf("Failed to send the TPM command to fd %d, got errno %d ="
+               "%s\n", ctx->tcpCtx.fd, errno, strerror(errno));
+    }
+#endif
+
+    return rc;
+}
+
+static TPM_RC SwTpmReceive(TPM2_CTX* ctx, void* buffer, size_t rxSz)
+{
+    TPM_RC rc = TPM_RC_SUCCESS;
+    ssize_t wrc = 0;
+    size_t bytes_remaining = rxSz;
+    char* ptr = buffer;
+
+    if (ctx == NULL || ctx->tcpCtx.fd <= 0 || buffer == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    while (bytes_remaining > 0) {
+        wrc = read(ctx->tcpCtx.fd, ptr, bytes_remaining);
+        if (wrc <= 0) {
+            #ifdef DEBUG_WOLFTPM
+            if (wrc == 0) {
+                printf("Failed to read from TPM socket: EOF\n");
+            } else {
+                printf("Failed to read from TPM socket %d, got errno %d"
+                       " = %s\n", ctx->tcpCtx.fd, errno, strerror(errno));
+            }
+            #endif
+            rc = SOCKET_ERROR_E;
+            break;
+        }
+
+        bytes_remaining -= wrc;
+        ptr += wrc;
+
+        #ifdef WOLFTPM_DEBUG_VERBOSE
+        printf("TPM socket received %zd waiting for %zu more\n",
+               wrc, bytes_remaining);
+        #endif
+    }
+
+    return rc;
+}
+
+static TPM_RC SwTpmConnect(TPM2_CTX* ctx, const char* host, const char* port)
+{
+    TPM_RC rc = SOCKET_ERROR_E;
+    struct addrinfo hints;
+    struct addrinfo *result, *rp;
+    int s;
+    int fd;
+
+    if (ctx == NULL) {
+        return BAD_FUNC_ARG;
+    }
+
+    memset(&hints, 0, sizeof(struct addrinfo));
```

Comment:
Let's change these to XMEMSET for portability, although I do realize this file is not portable and only designed for Linux.